tools/github_update.py

```diff
@@ -96,118 +105,237 @@ def _get_repo_path(self):
+    details_url = os.environ.get('GITHUB_COMMIT_STATUS_URL', '') # custom URL via env
+    if not details_url:
+        details_url = os.environ.get('BUILD_URL', '') # provided by jenkins
+        if details_url:
+            details_url += 'console'
+    return details_url
+
+
+def _should_access_github():
+    if 'WORKSPACE' not in os.environ:
+        # not running in CI. skip actually sending anything to GitHub
+        return False
+    if os.environ.get('GITHUB_DISABLE', ''):
+        # environment has notifications disabled. skip actually sending anything to GitHub
+        return False
+    if not (os.environ.get('GITHUB_COMMIT_STATUS_URL') or os.environ.get('BUILD_URL')):
+        # CI job was not triggered by a PR
+        return False
+    return True
 
 
 def print_help(argv):
-    logger.info('Syntax: {} <state: pending|success|error|failure> <context_label> <status message>'.format(argv[0]))
+    logger.info('Syntax:')
+    logger.info('- Update: {} <state: {}> <context_label> [a status message ...]'.format(
+        argv[0], '|'.join(VALID_STATES)))
+    logger.info('- Reset pending: {} reset [a replacement message ...]'.format(argv[0]))
+
+
+def reset_states(updater, message):
+    if not _should_access_github():
+        # reset disabled due to local build, exit silently
+        return 0
+
+    contexts = [context for context in updater.list_contexts() if context not in BLACKLISTED_CONTEXT_LABELS]
+    if not contexts:
+        # nothing to reset, exit silently
+        return 0
+
+    contexts.sort()
+    logger.info('[GH-STATUS] Resetting {} statuses: {}'.format(len(contexts), ', '.join(contexts)))
+    for context_label in contexts:
+        updater.update(PENDING_STATE, message=message, context_label=context_label)
+    return 0
+
+
+def set_state(updater, state, context_label, message):
+    if context_label in BLACKLISTED_CONTEXT_LABELS:
+        logger.error('Requested context label is on the blacklisted list: {}'.format(BLACKLISTED_CONTEXT_LABELS))
+        return 1
+
+    details_url = _get_details_link_url()
+    if details_url:
+        logmsg = '{} {}: {} ({})'.format(context_label, state, message, details_url)
+    else:
+        logmsg = '{} {}: {}'.format(context_label, state, message)
+
+    if _should_access_github():
+        logger.info('[GH-STATUS] {}'.format(logmsg))
+        updater.update(state, message=message, context_label=context_label, details_url=details_url)
+    else:
+        logger.info('[STATUS] {}'.format(logmsg))
+    return 0
 
 
 def main(argv):
-    if len(argv) < 4:
+    if len(argv) < 2:
         print_help(argv)
         return 1
-    state = argv[1]
-    if state != 'pending' \
-            and state != 'success' \
-            and state != 'error' \
-            and state != 'failure':
+
+    updater = GithubStatusUpdater()
+    command = argv[1]
+    if command == 'reset':
+        return reset_states(updater, ' '.join(argv[2:]))
```

Comment:
Felt like a nice to have, in case we wanted to say 'status reset' or similar in the cleared status. Didn't add much complexity.